File: medium/cpp/0133_clone_graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;

    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }

    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }

    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

#include <map>

class Solution {

public:
    void extract_graph(
        std::map<int, vector<int>> graph,
        Node* node
    ) {
        vector<int> target;
        graph[node->val] = target;
    }

public:
    Node* build_graph() {
        Node* new_node = new Node();
        return new_node;
    }

public:
    Node* cloneGraph(Node* node) {
        /* return NULL for empty graph */
        if (node == NULL) {
            return NULL;
        }

        /* copy graph if not empty */
        std::map<int, vector<int>> graph;
        Node* new_node = new Node();
        return new_node;
    }
};

File: medium/python3/0133_clone_graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:

    def extract_graph(self, graph: dict, node: 'Node'):
        graph[node.val] = []
        for n_ in node.neighbors:
            graph[node.val].append(n_.val)
            if n_.val not in graph:
                self.extract_graph(graph, n_)

    def build_graph(self, graph: dict) -> 'Node':
        nodes = {k_: Node(k_) for k_ in graph}
        for k_ in graph:
            for neighbor_ in graph[k_]:
                nodes[k_].neighbors.append(nodes[neighbor_])
        return nodes

    def cloneGraph(self, node: 'Node') -> 'Node':
        if node is None:
            return None
        graph = {}
        self.extract_graph(graph, node)
        nodes = self.build_graph(graph)
        return nodes[1]
